apps/prairielearn/assets/scripts/question.ts

```diff
@@ -15,349 +16,410 @@ import { setupCountdown } from './lib/countdown.js';
 import { mathjaxTypeset } from './lib/mathjax.js';
 
 onDocumentReady(() => {
-  const questionContainer = document.querySelector<HTMLElement>('.question-container');
-  if (questionContainer?.dataset.gradingMethod === 'External') {
-    externalGradingLiveUpdate();
-  }
+  // We use `selector-observer` here to handle the case of updating the page's
+  // contents without reloading the whole page. At the time of writing, this was
+  // used on the AI question generation draft editor page.
+  observe('.question-container', {
+    constructor: HTMLDivElement,
+    initialize(container) {
+      const controller = new QuestionContainerController(container);
+      return { remove: () => controller.cleanup() };
+    },
+  });
+});
 
-  const questionForm = document.querySelector<HTMLFormElement>('form.question-form');
-  if (questionForm) {
-    confirmOnUnload(questionForm);
-  }
+class QuestionContainerController {
+  private socket: Socket | null = null;
+  private abortController = new AbortController();
+  private submissionPanelObserver: Observer | null = null;
 
-  const markdownBody = document.querySelector<HTMLDivElement>('.markdown-body');
-  const revealFade = document.querySelector<HTMLDivElement>('.reveal-fade');
-  const expandButtonContainer = document.querySelector('.js-expand-button-container');
-  const expandButton = expandButtonContainer?.querySelector('button');
+  constructor(private container: HTMLElement) {
+    // TODO: is this the correct sequencing of MathJax?
+    void mathjaxTypeset([this.container]);
```

Comment:
This typeset call is a new addition. Normally this would be done on page load by MathJax (and it still is) - this just handles the case where we replace the question without reloading the page.

Re: sequencing, I think this is fine, but I left the TODO to call this out for reviewers. I'll remove it before merging.

Reply:
I don't see an issue with this sequencing, since none of the calls after this point affect the contents of any element that may have math. I would appreciate a comment here about the fact that this function is async and you're purposely not awaiting on it.